design/healcheck-endpoint.md

```diff
@@ -0,0 +1,169 @@
+# Healthcheck endpoint
+
+* **Status**: Draft #1
+* **JIRA**: [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+* **Observability JIRA**: [KEYCLOAK-8288](https://issues.jboss.org/browse/KEYCLOAK-8288)
+
+## Motivation
+
+Deploying both Keycloak and RHSSO on top of Kubernetes requires defining both Readiness Check and Liveness Check
+(see the [manual](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)).
+Currently, for Keycloak we use a very naive implementation, that calls `/auth/realms/master` URL.
+One of the biggest disadvantages of this approach is that it doesn't check the database state. The newly designed
+Healthcheck API presented in this design will address this problem.
+
+## Scope
+
+The scope of this Design includes:
+- Healthcheck API design
+- Implementation suggestions
+
+After many discussions, we decided to remove from the initial design the following items:
+- External IdP and User Federation checks. Including any of those checks makes the end-user situation worse.
+If we include an External IdP check in the Health endpoint, Kubernetes will plug all the pods out of the Service
+(a load balancer) if that IdP is down. Once all pods are removed from the Service, an end-user will see an HTTP 503
+Error page generated by the OpenShift router. If however, the External IdP is down, and we don't include it in the
+Health endpoint, a user will probably face an error during log in. But the error page will be displayed by Keycloak,
+and we control what message we send there. I consider this as a slightly better situation, although the end-result is
+the same - a user cannot log in.
+- Parametrized checks, e.g. `http://localhost:8080/auth/health?timeout=2s&fail-on=db,idp-google`. After removing
+the External IdPs and User Federation, the parametrized check is no longer necessary. Perhaps, we'll get back
+to it in the future.
+
+## Healthcheck parameters
+
+The new Health Endpoint need to support the `timeout` query parameter in the endpoint in the following form:
```

Comment:
Timeout parameter doesn't make sense to me. What would it actually control?

Reply:
To some degree I agree with you. If we rely solely on passive checks, we might remove it. However, if we ever do any form of active check (e.g. send ping to a database), we to supply some timeout for that operation. 

I guess we can remove it. It should be pretty straightforward to add it if needed.